thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/DetectionJobResource.java

```diff
@@ -307,31 +308,35 @@ public Response generateAnomaliesInRangeForFunctions(@QueryParam("ids") @NotNull
       LOG.warn("[Backfill] End time is in the future. Force to now.");
     }
 
-    final DateTime innerStartTime = startTime;
-    final DateTime innerEndTime = endTime;
+    final Map<Long, Integer> originalWindowSize = new HashMap<>();
+    final Map<Long, TimeUnit> originalWindowUnit = new HashMap<>();
+    final Map<Long, String> originalCron = new HashMap<>();
+    saveFunctionWindow(functionIdList, originalWindowSize, originalWindowUnit, originalCron);
+
+    // Update speed-up window and cron
+    if (speedup) {
+      for (long functionId : functionIdList) {
+        anomalyFunctionSpeedup(functionId);
+      }
+    }
+    // Run backfill
+    for (long functionId : functionIdList) {
+      long detectionJobId = detectionJobScheduler.runBackfill(functionId, startTime, endTime, forceBackfill);
+      detectionJobIdList.add(detectionJobId);
+    }
+
     new Thread(new Runnable() {
       @Override
       public void run() {
-        Map<Long, Integer> originalWindowSize = new HashMap<>();
-        Map<Long, TimeUnit> originalWindowUnit = new HashMap<>();
-        Map<Long, String> originalCron = new HashMap<>();
-        saveFunctionWindow(functionIdList, originalWindowSize, originalWindowUnit, originalCron);
-
-        // Update speed-up window and cron
-        if (speedup) {
-          for (long functionId : functionIdList) {
-            anomalyFunctionSpeedup(functionId);
-          }
+        for(long detectionJobId : detectionJobIdList) {
+          detectionJobScheduler.waitForJobDone(detectionJobId);
         }
-        // Run backfill
-        detectionJobScheduler.synchronousBackFill(functionIdList, innerStartTime, innerEndTime, forceBackfill);
-
         // Revert window setup
         revertFunctionWindow(functionIdList, originalWindowSize, originalWindowUnit, originalCron);
       }
     }).run();
 
-    return Response.ok().build();
+    return Response.ok(detectionJobIdList).build();
```

Comment:
Done. The return message now contains a map, which map the function id to its job execution id.